### PR TITLE
ROX-11931: Convert junit failure artifacts to Slack attachments

### PR DESCRIPTION
**Pull request by BradLugo**
## Description

Converts Junit artifacts that contain failures and failure messages to Slack message attachments.

Related to https://github.com/stackrox/junit-parse/pull/1

## Testing Performed

- Test CI: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/stackrox_stackrox/3438/pull-ci-stackrox-stackrox-master-gke-upgrade-tests/1588220418928414720 and successful Slack messages:https://srox.slack.com/archives/C740BUV25/p1667500352946239
- Tested with various artifact directories from failed Prow jobs

### CODE REVIEW

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1385]: this might need some extended bash globbing to get to junit xml in sub dirs. e.g. with this one:
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/branch-ci-stackrox-stackrox-master-merge-local-roxctl-tests/1582096454128766976
there is a report.xml in 
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/branch-ci-stackrox-stackrox-master-merge-local-roxctl-tests/1582096454128766976/artifacts/merge-local-roxctl-tests/stackrox-initial/artifacts/junit-reports/
I don't think that `find` will find it.

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1294]:
```suggestion
    if [[ -n "${slack_attachments}" ]]; then
```

**Review by gavin-stackrox — COMMENTED** [review]
In general, LGTM!, let me know when you are ready for a final review.

File: scripts/ci/lib.sh
Comment (BradLugo) [reply scripts/ci/lib.sh:1385]: This actually brings up an interesting point (along with your suggestion on how to test these changes); at least from what it seems in the failed job [#1582162936795762688](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/branch-ci-stackrox-stackrox-master-merge-local-roxctl-tests/1582162936795762688), spyglass only looks at `junit_operator.xml`. In fact, when I tried to parse `merge-local-roxctl-tests/stackrox-initial/artifacts/junit-reports/report.xml`, it threw an error which leads me to believe we're not properly formatting some of these reports. That said, I haven't dug into that error yet. It could actually be the case that spyglass isn't surfacing useful error messages.

So then, should we only look at `report.xml` if present for now? Or spend the time to debug these (seemingly) malformed junit reports and try to parse all XML files? Interested in your take on this.

Comment (gavin-stackrox) [reply scripts/ci/lib.sh:1385]: > So then, should we only look at report.xml if present for now?

Different jobs/tests name the JUnit differently. Sometimes single files, sometimes directories of XML (or maybe a directory with a single file). There is a wrapper script `store_test_results` to give you an idea of what you are dealing with.

[#1582162936795762688](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/branch-ci-stackrox-stackrox-master-merge-local-roxctl-tests/1582162936795762688) is a special but interesting case. The prow/spyglass Go code must be hitting the same error that you are and that is why it is not displayed in the UI. For errors we should do the same, silently drop them from the slack message. But it would be good to look at fixing the XML when it is dropped and/or fixing the parser in slack2junit and/or prow.

There are quite a few cases where you'll see prow/spyglass parse the Junit we (stackrox tests) create OK: 
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/branch-ci-stackrox-stackrox-master-merge-gke-qa-e2e-tests/1582357585149825024
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/branch-ci-stackrox-stackrox-master-merge-openshift-penultimate-qa-e2e-tests/1580969317434920960

Comment (BradLugo) [reply scripts/ci/lib.sh:1385]: Awesome, appreciate the feedback. I'll take a look at what's causing the malformed XML after this PR is merged.

---

Going back to the bash glob:
> this might need some extended bash globbing to get to junit xml in sub dirs

I thought `find` would do this automatically. From the POSIX manual:
> The find utility shall recursively descend the directory hierarchy from each file specified by path

Comment (gavin-stackrox) [reply scripts/ci/lib.sh]: not with GNU find 4.9.0:
```
$ mkdir -p ttt/junit-crud
$ touch ttt/junit-crud/report.xml
$ find ttt -type f -name 'junit-*.xml'
$
```

Comment (gavin-stackrox) [reply scripts/ci/lib.sh]: `**` will work though and does not seem to require `shopt extglob` (at least for my shell which is closer to the CI OS then say MacOS).
```
$ x=ttt/**/*.xml
$ echo $x
ttt/junit-crud/report-2.xml ttt/junit-crud/report.xml
$
```

**Comment by BradLugo** [comment]
@gavin-stackrox It seems that the unit test file is causing the CI to break. It seems to be breaking from the `| go list` pipe in the `go-unit-tests` make target. I think it's looking for a Go module from the master branch, but it's not there yet. Any thoughts or ideas on how to resolve this? I can ask the other engineers as well

File: scripts/ci/junit2slack/main.go
Comment (RTann) [inline scripts/ci/junit2slack/main.go:3]: the imports are out of order, which I think `style-checks` caught. You can run `make style` or `quickstyle`, if you have the https://github.com/stackrox/workflow scripts

File: scripts/ci/junit2slack/main.go
Comment (RTann) [inline scripts/ci/junit2slack/main.go:38]: one annoying thing about `Printf` if it does not add `\n`, so we might want to add that ourselves

File: scripts/ci/junit2slack/main.go
Comment (RTann) [inline scripts/ci/junit2slack/main.go:90]: you can just check `== 0`

File: scripts/ci/junit2slack/main.go
Comment (RTann) [inline scripts/ci/junit2slack/main.go:59]: you can also do `junitFiles ...*junit.Suites`. This way when you test with just one file, you can just do `convertJunitToSlack(suites)` where `suites` is of type `*junit.Suites`

File: scripts/ci/junit2slack/main.go
Comment (RTann) [inline scripts/ci/junit2slack/main.go:135]: you can just check `len(failedTestsBlocks) == 0`

File: scripts/ci/junit2slack/main.go
Comment (BradLugo) [reply scripts/ci/junit2slack/main.go:38]: Looks like each `Logger` method will end up using [Logger.Output](https://pkg.go.dev/log#Logger.Output) which says:

> A newline is appended if the last character of s is not already a newline.

That being said, I'm down to change this if we do something similar elsewhere in the repo

File: scripts/ci/junit2slack/main_test.go
Comment (BradLugo) [reply scripts/ci/junit2slack/main_test.go]: When I tried embedding the json file, it would add an extra `.` character to the end causing the unit test to fail. I'm not quite sure what was going on there

Comment (BradLugo) [reply scripts/ci/junit2slack/main_test.go]: Turns out it was a newline character

File: scripts/ci/junit2slack/main.go
Comment (misberner) [reply scripts/ci/junit2slack/main.go]: Generally I would prefer `failureMessage == ""` in Go. This is the most direct and has no performance disadvantages over `len(...) == 0`

**Comment by BradLugo** [comment]
Created a new repo to host the go project here: https://github.com/stackrox/junit-parse
Related PR: https://github.com/stackrox/junit-parse/pull/1

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1419]: Installing in CI can often lead to flakiness. It would be best to install this in the CI image to avoid the `go install` (stackrox/rox-ci-image:stackrox-test). However getting that image updated in CI is a major pain so I would rely on a future update to get the binary present. In the meantime you can `command -v` to check if it is present before installing. Something like: https://github.com/stackrox/stackrox/blob/d0cc2801cdd5373f6256d77aebe69f8cd83ae0c9/tests/upgrade/run.sh#L105

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1431]: Is this needed anymore? And I expect the dir is no longer present now that you have moved the go code.

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1433]: I'd suggest ` || return` following both of these. If there are any errors getting the tool or parsing the binary the plain slack message should still be sent.

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1434]: also delete?

**Review by gavin-stackrox — CHANGES_REQUESTED** [review]

File: scripts/ci/lib.sh
Comment (BradLugo) [reply scripts/ci/lib.sh]: Circling back to this since I ended up going with `find`; it looks like your example is looking for `junit-*.xml` but the name of the file is `report.xml`.

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1322]: 👍 to whatever works, we can iterate on it.

**Review by gavin-stackrox — COMMENTED** [review]
I'm ready to approve once you are done and all the forced failure & test changes are removed.

**Comment by BradLugo** [comment]
Sorry about that - I thought I had squared all that away before re-requesting your review. Should be good to go now

File: scripts/ci/lib.sh
Comment (gavin-stackrox) [inline scripts/ci/lib.sh:1323]: I think to handle the case when junit-parse is already in the test image:
```suggestion
            get_junit_parse_cli || true
        fi
        if command -v junit-parse >/dev/null 2>&1; then
```
We'll make a best effort try to get junit-parse and also handle when it is already present and not exit/error in either case.

**Review by gavin-stackrox — CHANGES_REQUESTED** [review]

**Review by gavin-stackrox — APPROVED** [review]
LGTM!

**Comment by BradLugo** [comment]
Thanks for all the help everyone!

**Pull request merged**
